pypose/module/constraint.py

```diff
@@ -0,0 +1,197 @@
+        # self.jacargs = {'vectorize':False, 'strategy':'reverse-mode'}
+
+    def forward(self, state, input):
+        r'''
+        Defines the computation.
+
+        '''
+        self.state, self.input = torch.atleast_1d(state), torch.atleast_1d(input)
+        return self.constraint(self.state, self.input)
+
+    def constraint(self, state, input, t=None):
+        r'''
+        Args:
+            state (:obj:`Tensor`): The state of the dynamical system
+            input (:obj:`Tensor`): The input to the dynamical system
+
+        Returns:
+            Tensor: The stage constraint of the system
+        '''
+        raise NotImplementedError("The users need to define their own state transition method")
+
+    def set_refpoint(self, state=None, input=None):
+        r'''
+        Function to set the reference point for linearization.
+
+        Args:
+            state (:obj:`Tensor`): The reference state of the dynamical system. If ``None``,
+                the the most recent state is taken. Default: ``None``.
+            input (:obj:`Tensor`): The reference input to the dynamical system. If ``None``,
+                the the most recent input is taken. Default: ``None``.
+
+        Returns:
+            None
+
+        Warning:
+            For nonlinear systems, the users have to call this function before getting the
+            linearized system.
+        '''
+        self._ref_state = torch.tensor(self.state) if state is None else torch.atleast_1d(state)
+        self._ref_input = torch.tensor(self.input) if input is None else torch.atleast_1d(input)
+        self._ref_c = self.constraint(self._ref_state, self._ref_input)
+
+    @property
+    def gx(self):
+        r'''
+        Linear/linearized constraint term on state
+
+        .. math::
+            \mathbf{g}_{\mathbf_{x}} = \left. \frac{\partial \mathbf{g}}{\partial \mathbf{x}} \right|_{\chi^*}
+        '''
+        func = lambda x: self.constraint(x, self._ref_input)
+        return excludeBatch(jacobian(func, self._ref_state, **self.jacargs), type=2)
+
+    @property
+    def gu(self):
+        r'''
+        Linear/linearized constraint term on input
+
+        .. math::
+            \mathbf{g}_{\mathbf_{u}} = \left. \frac{\partial \mathbf{g}}{\partial \mathbf{u}} \right|_{\chi^*}
+        '''
+        func = lambda u: self.constraint(self._ref_state, u)
+        return excludeBatch(jacobian(func, self._ref_input, **self.jacargs), type=2)
+
+    @property
+    def g(self):
+        r'''
+        Constant term generated by constraint.
+
+        .. math::
+            \mathbf{g} = \mathrm{cons}(\mathbf{x}^*, \mathbf{u}^*, t^*)
+                           - \mathbf{g}_{\mathbf{x}}\mathbf{x}^* - \mathbf{g}_{\mathbf{u}}\mathbf{u}^*
+        '''
+        # Potential performance loss here - self.A and self.B involves jacobian eval
+        return self._ref_c - pp.bmv(self.gx, self._ref_state) - pp.bmv(self.gu, self._ref_input)
+
+class LinCon(Constraint):
```

Comment:
Is this constraint inequality constraint? Didn't see anywhere in the doc defines that and whether it's >0, or <0, or >=0, or <=0.

Reply:
Added to the docs in constraint.py